### ISSUE TITLE
Selected certificate without chain

### ISSUE THREAD

**Issue by crashodd** [FlorianSW/org.droidwiki.certtest#2]
Hey Florian! nice to talk to you.

I ended up with your example which is quite a great solution for what we are looking for.
Our use case would be to select a certificate and with it make a call to a ssl url.

this is kind of what we want to do:
```
ks = KeyStore.getInstance("PKCS12");
ks.load(null, null);
ks.setCertificateEntry(certificateAlias, cert);

TrustManagerFactory tmf = TrustManagerFactory.getInstance("SunX509");
tmf.init(ks);

// Create an SSLContext that uses our TrustManager
SSLContext context = SSLContext.getInstance("TLS");
context.init(null, tmf.getTrustManagers(), null);
```
An proceed to the call with the ssl context. Although I have imported root and middle certificate, is not working saying 

> No trusted certificate found

Using jvm option to debug ssl, I can see that when selecting the certificate is not loading the certificate chain, which I guess that latter will be used to validate the one requested from the server.
With this chain I would add it as an entry to the created keystore.

Is there any way to fetch this information in an easy way? maybe extending CERT_EXTENSION structure?

**Comment by tresf** [comment]
@crashodd perhaps you're hoping that Java would [trust the Windows certificate store](https://stackoverflow.com/a/52330588/3196753)?

Otherwise, you'd have to [install the certificate into the Java trusted store](https://stackoverflow.com/a/18889855/3196753)

Lastly, if you want to write your own trust handler in the connector there [are ways to do that as well](https://stackoverflow.com/q/2642777/3196753).

In summary, your question appears to be more around "Why doesn't Java trust my Windows Cert Store", which (I believe) this repository does not aim to achieve.

Disclaimer, I'm not the maintainer of this repository, just a person who's found it useful. :D

**Comment by crashodd** [comment]
@tresf thanks for your answer, but basically the flow would be to provide users with certificates, and this desktop application would prompt for the installed certs. 

My problem is that I always came up with a private key that is a wrapper, basically cannot be done in java with mscapi as also stated [here](https://stackoverflow.com/questions/22332280/convert-private-key-to-pkcs8-format-in-java)
Probably using jna or so it can be done but might not be worthit.

Anyways thanks again for the response!

**Comment by tresf** [comment]
@crashodd the Java BouncyCastle library can convert to and from just about any private key and/or certificate format.   I'm confused as the 6-year old article that is linked doesn't make much sense in 2020.

I took a liking to @FlorianSW's project specifically for the well documented JNA hooks to the Windows store, but I don't think Java will trust custom Windows-installed certificates unless you install them into Java and Java generally supports `.jks` and `.pfx`.  `.pfx` is now encouraged because it's not a proprietary format... anyway, the code to install into the Java store usually requires dealing with a store that's already on your system, so you'd just need to get the `x509` cert and install it using the link in my previous post.

Furthermore, in regards to Windows, there are other ways to install a certificate to the system other than raising this dialog.  It's not well documented, so [here's a code example](https://github.com/tresf/tray/blob/7ede43f5710005a748f8719199cb95a702b46989/src/qz/installer/certificate/WindowsCertificateInstaller.java#L101-L107).

**Comment by tresf** [comment]
@crashodd it looks like another trick is to use a system property to tell it to trust your store as the cert store, also mentioned in the article linked previously.
```java
System.setProperty("javax.net.ssl.trustStore", fileName);
```

**Comment by FlorianSW** [comment]
> @tresf thanks for your answer, but basically the flow would be to provide users with certificates, and this desktop application would prompt for the installed certs.
> 
> My problem is that I always came up with a private key that is a wrapper, basically cannot be done in java with mscapi as also stated [here](https://stackoverflow.com/questions/22332280/convert-private-key-to-pkcs8-format-in-java)
> Probably using jna or so it can be done but might not be worthit.
> 
> Anyways thanks again for the response!

That sounds scary and as I know of, the KeyStore returned from "WINDOWS-MY" (with the `SunMSCAPI`, which you probably use) does not support to retrieve the private key of a key entry in an encoded format for security reasons.

By simply googling for that, I think a better way would be something like this:
https://stackoverflow.com/questions/52194306/using-javas-sunmscapi-windows-my-to-access-smartcard-certificates-for-tls-ssl-c

I might've misunderstood what you actually want to do, though.
